### PR TITLE
Feature/form validation

### PR DESCRIPTION
**Pull request by andrupu**
PXBLUE-1318

Changes proposed in this Pull Request:

- Create the form validation example based on the existing angular design patterns example

### CODE REVIEW

**Comment by emclaug2** [comment]
To have this tested on a mobile device, you need to add the Form Validation navigation item to the Drawer.  Otherwise can only be tested on a browser which isn't the targeted environment for these ionic apps.

**Comment by joebochill** [comment]
I'm unable to run the dev branch on a mobile device (at least on my new machine). This was the main reason for the discussion we had last week about the need to re-scaffold this repository. Seems it's gotten a little out of date since we were last in here. I'll check and see if it will run on my old mac.

**Comment by daileytj** [comment]
I'm not having much luck running this in a browser either.
Screen Shot 2021-05-03 at 8 49 17 AM

**Comment by joebochill** [comment]
> I'm unable to run the dev branch on a mobile device (at least on my new machine). This was the main reason for the discussion we had last week about the need to re-scaffold this repository. Seems it's gotten a little out of date since we were last in here. I'll check and see if it will run on my old mac.

No dice on the old mac...my ionic environment over there is goofy.

@andrupu if you want to do the rescaffolding as part of this PR, that would be helpful. You would need to generate a new ionic project using the PX Blue CLI and then copy over the 'guts' of the existing project.

```sh
npx -p @pxblue/cli pxb new ionic --name=ionic-design-patterns
```

**Comment by andrupu** [comment]
I can't seem to create a new ionic scaffold.  It keeps giving me this error:
image

**Comment by joebochill** [comment]
I just ran through and created a new project on my mac and it worked as expected. I don't have a PC to test that, but maybe somebody else can confirm that.

Might be related to: 
image

Make sure you are creating the new project in a new folder, not inside of the ionic-design-patterns repository that you have cloned down.

**Comment by andrupu** [comment]
I figured it out... I had to set NODE_TLS_REJECT_UNAUTHORIZED=0 due to the Eaton proxy in order to create a new Ionic scaffold.

**Comment by joebochill** [comment]
> I figured it out... I had to set NODE_TLS_REJECT_UNAUTHORIZED=0 due to the Eaton proxy in order to create a new Ionic scaffold.

Make sure you unset that when you are done :)

**Comment by emclaug2** [comment]
It looks like the build is still failing; maybe remove the yarn.lock and create a new one by running `yarn`?

File: .browserslistrc [hunk 1]
```diff
@@ -0,0 +1,17 @@
+# This file is used by the build system to adjust CSS and JS output to support the specified browsers below.
```
Comment (joebochill) [inline .browserslistrc:1]: We should use our standard browsers list browsers in here (refer to the old file)

File: README.md [hunk 2]
```diff
@@ -13,6 +13,7 @@ This repository holds the source code examples for [PX Blue design patterns](htt
...
 | [Data List](https://pxblue.github.io/patterns/lists)                           | A basic list dynamically built from a JSON data object.                                                                                                                                      | [StackBlitz](https://stackblitz.com/github/pxblue/ionic-design-patterns?file=src%2Fapp%2Fpages%2Flist%2Fdata-list%2Fdata-list.component.html)                               |
+| [Form Validation](https://pxblue.github.io/patterns/lists)                     | Form Fields and various imput fields for validation usage                                                                                                                                    | [StackBlitz]()                               |
```
Comment (joebochill) [inline README.md:16]: Use the same description here from the other pattern repos: 'Various types of form validation, such as verifying phone number, email address, character length, and password.'

File: .eslintrc.json [hunk 3]
```diff
@@ -0,0 +1,47 @@
+{
```
Comment (joebochill) [inline .eslintrc.json:1]: Did you select the PX Blue lint/prettier options when you rescaffolded this? This doesn't look like our typical lint config.

File: angular.json [hunk 4]
```diff
@@ -14,25 +14,24 @@
...
+                "input": "src/assets",
+                "output": "assets"
+              },
+              {
+                "glob": "**/*.svg",
+                "input": "node_modules/ionicons/dist/ionicons/svg",
+                "output": "./svg"
+              }
             ],
+            "styles": ["src/theme/variables.scss", "src/global.scss"],
```
Comment (joebochill) [inline angular.json:34]: You are missing all of the px blue theming styles in here...are you sure you used our CLI?

File: e2e/src/app.e2e-spec.ts [hunk 5]
```diff
@@ -1,23 +1,14 @@
...
-    expect(page.getTitleText()).toEqual('angular-design-patterns app is running!');
-  });
-
-  afterEach(async () => {
-    // Assert that there are no errors emitted from the browser
-    const logs = await browser.manage().logs().get(logging.Type.BROWSER);
-    expect(logs).not.toContain(jasmine.objectContaining({
-      level: logging.Level.SEVERE,
-    } as logging.Entry));
+    expect(page.getParagraphText()).toContain('Start with Ionic UI Components');
```
Comment (joebochill) [inline e2e/src/app.e2e-spec.ts:12]: We should have tests in here similar to the tests in the other pattern repositories (see the old file).

File: e2e/tsconfig.json [hunk 6]
```diff
@@ -3,10 +3,9 @@
   "compilerOptions": {
     "outDir": "../out-tsc/e2e",
     "module": "commonjs",
-    "target": "es5",
+    "target": "es2018",
```
Comment (joebochill) [inline e2e/tsconfig.json:6]: I think we may want to stick with es5 here

File: ionic.starter.json [hunk 7]
```diff
@@ -0,0 +1,12 @@
+{
```
Comment (joebochill) [inline ionic.starter.json:1]: What is this file (is this a new ionic thing)?

File: package.json [hunk 8]
```diff
@@ -1,103 +1,60 @@
 {
   "name": "ionic-design-patterns",
-  "version": "0.0.0",
+  "version": "0.0.1",
+  "author": "Ionic Framework",
+  "homepage": "https://ionicframework.com/",
```
Comment (joebochill) [inline package.json:5]: This should use the PX Blue author information

File: package.json [hunk 9]
```diff
@@ -1,103 +1,60 @@
...
     "start": "ng serve",
     "build": "ng build",
     "test": "ng test",
-    "test:ci": "ng test --watch false",
-    "lint": "eslint \"src/**/**.ts\"",
-    "e2e": "ng e2e",
-    "lint:fix": "eslint \"src/**/**.ts\" --fix",
-    "prettier": "prettier \"src/**/**.{ts,js,json,css,scss,html}\" --write",
-    "prettier:check": "prettier \"src/**/**.{ts,tsx,js,jsx,json,css,scss,html}\" --check"
+    "lint": "ng lint",
```
Comment (joebochill) [inline package.json:11]: Missing the PX Blue lint and prettier scripts.

File: package.json [hunk 10]
```diff
@@ -1,103 +1,60 @@
...
-    "core-js": "^2.5.4",
-    "node-sass": "^4.13.1",
-    "placeholder-loading": "^0.2.6",
+    "@angular/common": "~11.2.0",
+    "@angular/core": "~11.2.0",
+    "@angular/forms": "~11.2.0",
+    "@angular/platform-browser": "~11.2.0",
+    "@angular/platform-browser-dynamic": "~11.2.0",
+    "@angular/router": "~11.2.0",
+    "@ionic/angular": "^5.5.2",
```
Comment (joebochill) [inline package.json:22]: We're missing all of the PX Blue packages and all of the cordova plugins...I'm thinking you may have scaffolded this with the ionic cli and not the px blue cli...

**Review by joebochill — CHANGES_REQUESTED** [review]

File: ionic.starter.json (hunk 7 above)
Comment (andrupu) [reply ionic.starter.json:1]: Yes, this is a new generated file in Ionic since the last PXBlue CLI update.

**Comment by emclaug2** [comment]
@andrupu Did you push out your latest changes?  I cannot run the angular app, the ionic app doesn't work and the build is still passing, same error as last week.

**Comment by andrupu** [comment]
> @andrupu Did you push out your latest changes? I cannot run the angular app, the ionic app doesn't build either and the ci/cd pipeline is same error as last week.

Silly me.  I forgot to run yarn.  Also, I found and deleted the redundant browserlist file (browerlistrc) that Circle CI was throwing an error on.

**Comment by joebochill** [comment]
So I was able to get it to run, but in doing so it shows 60+ file changes.

image

Most of these are in the resources folder (make sure the gitignore file is updated so that these generated files are not included in the source control).

The config.xml also got overridden with a generic projct name (I suspect you just need to run the project for this file tog et generated and then you can change the strings for project name, description, etc.)

The Form Validation example itself looks off:

image

It looks like we are closer to getting the re-scaffolding working than the form stuff, so it might behoove us to split this PR (get the final rescaffolding stuff cleaned up and then make a separate PR with the form validation stuff).

The test step is still failing in CI:
image

This is something you can check locally by running `yarn test:ci` — it still fails even after adding the missing dependency, so looks like there's still a little work needed there.

**Comment by joebochill** [comment]
Are we going to close this PR and make a new one with the form validation example now that th scaffold has been updated in the other PR?

**Comment by andrupu** [comment]
> Are we going to close this PR and make a new one with the form validation example now that th scaffold has been updated in the other PR?

Yes, now that the Scaffold has been merged with the dev branch I am going to close this PR and create a fresh one branched from the updated dev branch.  I will take care of the other branches I created, I just need them to bring the Form Validation changes forward to the new PR.

**Pull request closed without merging**